### PR TITLE
Update proband dept

### PR DESCRIPTION
**Pull request by rkrenn**
when creating a proband, it is linked to the creating user's department. any record field such as first name, last name, DoB etc. holding PII (personally identifyable information) is encrypted using that departments key. the same applies to detail records (addresses, contact details, comments, ..) belonging to the proband.

once created, it was therefor not allowed to change a proband's department. to enable migrating a proband from one department to another, the re-encryption is required.

### CODE REVIEW

File: core/src/main/java/org/phoenixctms/ctsms/service/proband/ProbandServiceImpl.java [hunk 1]
```diff
@@ -2905,4 +2912,150 @@ private void notifyProbandInactive(ProbandStatusEntry statusEntry, Date now) thr
...
+		}
+		Department oldDepartment = proband.getDepartment();
+		if (!oldDepartment.equals(newDepartment)) {
+			if (!CryptoUtil.checkDepartmentPassword(newDepartment, plainNewDepartmentPassword)) {
+				throw L10nUtil.initServiceException(ServiceExceptionCodes.DEPARTMENT_PASSWORD_WRONG);
+			}
+			if (!CryptoUtil.checkDepartmentPassword(oldDepartment, plainOldDepartmentPassword)) {
+				throw L10nUtil.initServiceException(ServiceExceptionCodes.OLD_DEPARTMENT_PASSWORD_WRONG);
+			}
+			Timestamp now = new Timestamp(System.currentTimeMillis());
```
Comment (laborg) [inline core/src/main/java/org/phoenixctms/ctsms/service/proband/ProbandServiceImpl.java]: What happens if an exception occurs in one of the re-encryptors?

File: core/src/main/java/org/phoenixctms/ctsms/security/reencrypt/FileReEncrypter.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		InputStream plainTextStream = new VerifyMD5InputStream(
+				CryptoUtil.createDecryptionStream(getIv(item), oldDepartmentKey, CoreUtil.createFileServiceFileInputStream(getExternalFileName(item))), getMd5(item));
+		String newFileName = CoreUtil.createExternalFileName(CoreUtil.getExternalFileDirectoryPrefix(getModule(item)), getExternalFileName(item));
+		java.io.File externalFile = new java.io.File(CoreUtil.getFileServiceExternalFilename(newFileName));
+		CipherStream cipherStream = CryptoUtil.createEncryptionStream(newDepartmentKey, new FileOutputStream(externalFile));
+		int nRead;
+		long totalRead = 0;
+		byte[] block = new byte[CommonUtil.INPUTSTREAM_BUFFER_BLOCKSIZE];
+		try {
+			while ((nRead = plainTextStream.read(block, 0, block.length)) != -1) {
```
Comment (laborg) [inline core/src/main/java/org/phoenixctms/ctsms/security/reencrypt/FileReEncrypter.java]: IOUtils offers `copy` for this purpose.

File: core/src/main/java/org/phoenixctms/ctsms/service/proband/ProbandServiceImpl.java (hunk 1 above)
Comment (rkrenn) [reply core/src/main/java/org/phoenixctms/ctsms/service/proband/ProbandServiceImpl.java]: the transaction is abandoned instantly. this also holds for the re-encryption of large files.

File: core/src/main/java/org/phoenixctms/ctsms/security/reencrypt/FileReEncrypter.java (hunk 2 above)
Comment (rkrenn) [reply core/src/main/java/org/phoenixctms/ctsms/security/reencrypt/FileReEncrypter.java]: seems we have commons-io as dep already. but it's currently not used anywhere afaics

**Comment by laborg** [comment]
I'd cut roughly 90% of code by using this class:

```java
	private static class Recryptor {

		private SecretKey oldDepartmentKey;
		private SecretKey newDepartmentKey;

		public Recryptor(SecretKey oldDepartmentKey, SecretKey newDepartmentKey) {
			super();
			this.oldDepartmentKey = oldDepartmentKey;
			this.newDepartmentKey = newDepartmentKey;
		}

		private <T> void doField(T item, Function<T, byte[]> ivGetter, BiConsumer<T, byte[]> ivSetter, Function<T, byte[]> encryptedGetter,
				BiConsumer<T, byte[]> encryptedSetter, BiConsumer<T, byte[]> hashSetter) throws Exception {
			byte[] plainText = CryptoUtil.decrypt(ivGetter.apply(item), oldDepartmentKey, encryptedGetter.apply(item));
			CipherText cipherText = CryptoUtil.encrypt(newDepartmentKey, plainText);
			ivSetter.accept(item, cipherText.getIv());
			encryptedSetter.accept(item, cipherText.getCipherText());
			hashSetter.accept(item, CryptoUtil.hashForSearch(newDepartmentKey, plainText));
		}

		private <T> void doData(T item, Function<T, byte[]> ivGetter, BiConsumer<T, byte[]> ivSetter, Function<T, byte[]> encryptedGetter,
				BiConsumer<T, byte[]> encryptedSetter) throws Exception {
			byte[] plainText = CryptoUtil.decrypt(ivGetter.apply(item), oldDepartmentKey, encryptedGetter.apply(item));
			CipherText cipherText = CryptoUtil.encrypt(newDepartmentKey, plainText);
			ivSetter.accept(item, cipherText.getIv());
			encryptedSetter.accept(item, cipherText.getCipherText());
		}
		
		private <T> void doFile(File item) throws Exception {
			InputStream plainTextStream = new VerifyMD5InputStream(
					CryptoUtil.createDecryptionStream(item.getDataIv(), oldDepartmentKey, CoreUtil.createFileServiceFileInputStream(item.getExternalFileName())), item.getMd5());
			String newFileName = CoreUtil.createExternalFileName(CoreUtil.getExternalFileDirectoryPrefix(item.getModule()), item.getExternalFileName());
			java.io.File externalFile = new java.io.File(CoreUtil.getFileServiceExternalFilename(newFileName));
			CipherStream cipherStream = CryptoUtil.createEncryptionStream(newDepartmentKey, new FileOutputStream(externalFile));
			IOUtils.copy(plainTextStream, cipherStream);
			item.setExternalFileName(newFileName);
			item.setDataIv(cipherStream.getIv());
		}
	}
```

and then using it like this 

```java
	protected ProbandOutVO handleUpdateProbandDepartment(AuthenticationVO auth, Long probandId, Long newDepartmentId, String plainNewDepartmentPassword,
			String plainOldDepartmentPassword) throws Exception {

			// build up
			ProbandContactParticulars pp = proband.getPersonParticulars();
			if (proband.isPerson() && pp != null) {
				recryptor.doData(pp, ProbandContactParticulars::getPrefixedTitle1Iv, ProbandContactParticulars::setPrefixedTitle1Iv,
						ProbandContactParticulars::getEncryptedPrefixedTitle1, ProbandContactParticulars::setEncryptedPrefixedTitle1);
				recryptor.doData(pp, ProbandContactParticulars::getPrefixedTitle2Iv, ProbandContactParticulars::setPrefixedTitle2Iv,
						ProbandContactParticulars::getEncryptedPrefixedTitle2, ProbandContactParticulars::setEncryptedPrefixedTitle2);
				recryptor.doData(pp, ProbandContactParticulars::getPrefixedTitle3Iv, ProbandContactParticulars::setPrefixedTitle3Iv,
						ProbandContactParticulars::getEncryptedPrefixedTitle3, ProbandContactParticulars::setEncryptedPrefixedTitle3);
				recryptor.doField(pp, ProbandContactParticulars::getFirstNameIv, ProbandContactParticulars::setFirstNameIv,
						ProbandContactParticulars::getEncryptedFirstName, ProbandContactParticulars::setEncryptedFirstName, ProbandContactParticulars::setFirstNameHash);
				recryptor.doField(pp, ProbandContactParticulars::getLastNameIv, ProbandContactParticulars::setLastNameIv,
						ProbandContactParticulars::getEncryptedLastName, ProbandContactParticulars::setEncryptedLastName, ProbandContactParticulars::setLastNameHash);
						// add rest		
			} 
			// add rest
			ProbandAddressDao probandAddressDao = this.getProbandAddressDao();
			Iterator<ProbandAddress> addressesIt = proband.getAddresses().iterator();
			while (addressesIt.hasNext()) {
				ProbandAddress address = addressesIt.next();
				recryptor.doField(address, ProbandAddress::getCountryNameIv, ProbandAddress::setCountryNameIv,
						ProbandAddress::getEncryptedCountryName, ProbandAddress::setEncryptedCountryName, ProbandAddress::setCountryNameHash);
				recryptor.doField(address, ProbandAddress::getZipCodeIv, ProbandAddress::setZipCodeIv,
						ProbandAddress::getEncryptedZipCode, ProbandAddress::setEncryptedZipCode, ProbandAddress::setZipCodeHash);
				// add rest
				CoreUtil.modifyVersion(address, address.getVersion(), now, user);
				probandAddressDao.update(address);
			}
			//  add rest
			FileDao fileDao = this.getFileDao();
			Iterator<File> filesIt = proband.getFiles().iterator();
			while (filesIt.hasNext()) {
				File file = filesIt.next();
				recryptor.doFile(file);
				CoreUtil.modifyVersion(file, file.getVersion(), now, user);
				fileDao.update(file);
			}
```

**Comment by rkrenn** [comment]
the .doX() invocations should be wired in the dao then, as the central place. otherwise accessing fields is duplicated at each servicemethod that needs that logic. 

what about IOUtils, should we introduce commons.io dependency? would need to walk through all places that use the stream read-write pattern.

**Pull request merged**